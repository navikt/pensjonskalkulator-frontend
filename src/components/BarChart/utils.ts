import { ChartData } from './BarChart'

export const getBarChartHeight = (
  value: number,
  maxValue: number,
  maxHeight: number = 128
) => {
  if (value === 0) {
    return 0
  }
  return (value / maxValue) * maxHeight
}

export const findMaxValue = (data: ChartData[]): number => {
  return data.reduce((max, { value }) => (max > value ? max : value), 0)
}
