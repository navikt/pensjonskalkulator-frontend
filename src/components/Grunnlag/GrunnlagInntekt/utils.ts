export const validateInntektInput = (
  s: string | undefined,
  updateValidationErrorMessage: (s: string) => void
) => {
  let isValid = true
  if (s === undefined || s === '') {
    isValid = false
    updateValidationErrorMessage(
      'grunnlag.inntekt.inntektmodal.textfield.validation_error.required'
    )
  } else if (isNaN(s as unknown as number) || !/^[0-9]+$/.test(s)) {
    isValid = false
    updateValidationErrorMessage(
      'grunnlag.inntekt.inntektmodal.textfield.validation_error.type'
    )
  } else if (parseInt(s as string, 10) > 100000000) {
    isValid = false
    updateValidationErrorMessage(
      'grunnlag.inntekt.inntektmodal.textfield.validation_error.max'
    )
  }
  return isValid
}
