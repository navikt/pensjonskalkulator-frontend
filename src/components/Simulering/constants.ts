export const MAX_UTTAKSALDER = 77
export const COLUMN_WIDTH = 25
export const TOOLTIP_YPOS = 35

export const SERIES_COLORS = {
  SERIE_COLOR_INNTEKT: 'var(--a-gray-500)',
  SERIE_COLOR_AFP: 'var(--a-purple-400)',
  SERIE_COLOR_TP: 'var(--a-green-400)',
  SERIE_COLOR_ALDERSPENSJON: 'var(--a-deepblue-500)',
  SERIE_COLOR_FADED_INNTEKT: 'var(--a-gray-300)',
  SERIE_COLOR_FADED_AFP: 'var(--a-purple-200)',
  SERIE_COLOR_FADED_TP: 'var(--a-green-200)',
  SERIE_COLOR_FADED_ALDERSPENSJON: 'var(--a-deepblue-200)',
}

export const SERIES_DEFAULT = {
  SERIE_INNTEKT: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'Pensjonsgivende inntekt',
    color: SERIES_COLORS.SERIE_COLOR_INNTEKT,
  },
  SERIE_TP: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'Pensjonsavtaler (arbeidsgivere m.m.)',
    color: SERIES_COLORS.SERIE_COLOR_TP,
  },
  SERIE_AFP: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'AFP (Avtalefestet pensjon)',
    color: SERIES_COLORS.SERIE_COLOR_AFP,
  },
  SERIE_ALDERSPENSJON: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'Alderspensjon (NAV)',
    color: SERIES_COLORS.SERIE_COLOR_ALDERSPENSJON,
  },
}

export const highchartsScrollingSelector = '.highcharts-scrolling'
