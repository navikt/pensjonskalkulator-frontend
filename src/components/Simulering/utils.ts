import { IntlShape } from 'react-intl'

import { Chart, Point, Series } from 'highcharts'

import { getAlderMinus1Maaned } from '@/utils/alder'

import {
  COLUMN_WIDTH,
  MAX_UTTAKSALDER,
  SERIES_COLORS,
  highchartsScrollingSelector,
} from './constants'
import {
  ExtendedAxis,
  HighchartsScrollingHTMLDivElement,
} from './utils-highcharts'

export const getChartDefaults = (aarArray: string[]) => {
  return {
    chart: {
      type: 'column',
      scrollablePlotArea: {
        minWidth: aarArray.length * COLUMN_WIDTH * 1.6,
        scrollPositionX: 0,
      },
    },
    xAxis: {
      categories: aarArray,
    },
  }
}

export const processInntektArray = (args: {
  xAxisStartAar: number
  inntektFoerUttakBeloep: number
  gradertUttak:
    | {
        fra: Alder
        til: Alder
        beloep?: number
      }
    | undefined
  heltUttak:
    | {
        fra: Alder
        til?: Alder
        beloep?: number
      }
    | undefined

  xAxisLength: number
}): number[] => {
  const {
    xAxisStartAar,
    inntektFoerUttakBeloep,
    gradertUttak,
    heltUttak,
    xAxisLength,
  } = args

  const utbetalingsperioder = [
    // før uttak
    {
      startAlder: { aar: xAxisStartAar, maaneder: 0 },
      sluttAlder: gradertUttak
        ? getAlderMinus1Maaned(gradertUttak.fra)
        : heltUttak
          ? getAlderMinus1Maaned(heltUttak.fra)
          : undefined,
      aarligUtbetaling: inntektFoerUttakBeloep,
    },
    // gradert
    ...(gradertUttak
      ? [
          {
            startAlder: gradertUttak.fra,
            sluttAlder: gradertUttak.til,
            aarligUtbetaling: gradertUttak.beloep ?? 0,
          },
        ]
      : []),
    // helt
    ...(heltUttak
      ? [
          {
            startAlder: heltUttak.fra,
            sluttAlder: heltUttak.til,
            aarligUtbetaling: heltUttak.beloep ?? 0,
          },
        ]
      : []),
  ]

  const sluttAlder = xAxisStartAar + xAxisLength - 1
  const result = new Array(sluttAlder - xAxisStartAar + 1).fill(0)

  utbetalingsperioder.forEach((utbetalingsperiode) => {
    const periodeStartYear = Math.max(
      xAxisStartAar,
      utbetalingsperiode.startAlder.aar
    )
    const avtaleEndYear = utbetalingsperiode.sluttAlder
      ? Math.min(sluttAlder, utbetalingsperiode.sluttAlder.aar)
      : sluttAlder

    for (let year = periodeStartYear; year <= avtaleEndYear; year++) {
      if (year >= xAxisStartAar) {
        const antallMaanederMedPensjon = getAntallMaanederMedPensjon(
          year,
          utbetalingsperiode.startAlder,
          utbetalingsperiode.sluttAlder
        )

        const allocatedAmount =
          (utbetalingsperiode.aarligUtbetaling *
            Math.max(0, antallMaanederMedPensjon)) /
          12
        result[year - xAxisStartAar] += allocatedAmount
      }
    }
  })

  return result
}

export const processPensjonsberegningArray = (
  pensjonsberegninger: AfpPrivatPensjonsberegning[] = [],
  isEndring: boolean,
  xAxisLength: number
): number[] => {
  const arrayLength = Math.max(
    xAxisLength,
    isEndring ? pensjonsberegninger.length + 1 : pensjonsberegninger.length + 2
  )
  const dataArray = isEndring ? [] : new Array(1).fill(0)

  const livsvarigPensjonsbeloep =
    pensjonsberegninger[pensjonsberegninger.length - 1]?.beloep ?? 0

  for (let index = isEndring ? 0 : 1; index < arrayLength; index++) {
    const pensjonsBeregningAtIndex =
      pensjonsberegninger[isEndring ? index : index - 1]
    dataArray.push(
      pensjonsBeregningAtIndex
        ? pensjonsBeregningAtIndex.beloep
        : livsvarigPensjonsbeloep
    )
  }
  return dataArray
}

export const processAfpPensjonsberegningArray = (
  xAxisStartAar: number, // uttaksaar, (uttaksaar minus 1 for førstegangsøkere)
  xAxisLength: number,
  pensjonsberegninger: AfpPrivatPensjonsberegning[] = [],
  isEndring: boolean
): number[] => {
  if (pensjonsberegninger.length === 0) {
    return []
  }
  const arrayLength = Math.max(
    xAxisLength,
    isEndring ? pensjonsberegninger.length + 1 : pensjonsberegninger.length + 2
  )

  const startYear = pensjonsberegninger[0].alder
  const emptyYearsBeforeStart = startYear - xAxisStartAar
  const dataArray = isEndring ? [] : new Array(1).fill(0)
  const startIndex = emptyYearsBeforeStart ?? (isEndring ? 0 : 1)

  const livsvarigPensjonsbeloep =
    pensjonsberegninger[pensjonsberegninger.length - 1]?.beloep ?? 0

  for (let index = isEndring ? 0 : 1; index < arrayLength; index++) {
    if (startIndex > index) {
      dataArray.push(0)
    } else {
      const pensjonsBeregningAtIndex = pensjonsberegninger[index - startIndex]
      dataArray.push(
        pensjonsBeregningAtIndex
          ? pensjonsBeregningAtIndex.beloep
          : livsvarigPensjonsbeloep
      )
    }
  }
  return dataArray
}

// Antall maaneder i en avtale beregnes "Fra og med" og "Til og med"
export const getAntallMaanederMedPensjon = (
  year: number,
  utbetalingsperiodeStartAlder: Alder,
  utbetalingsperiodeSluttAlder?: Alder
) => {
  // Hvis vi viser første år av avtalen, tar vi høyde for startMaaned, hvis ikke teller avtalen fra måned 0 (fult år)
  const periodStartMonth =
    utbetalingsperiodeStartAlder.aar === year
      ? utbetalingsperiodeStartAlder.maaneder
      : 0

  // Hvis avtalen har en sluttdato og at vi viser siste av avtalen, tar vi høyde for sluttMaaned, hvis ikke teller avtalen til måned 11 (fult år)
  const periodEndMonth =
    utbetalingsperiodeSluttAlder && year === utbetalingsperiodeSluttAlder?.aar
      ? utbetalingsperiodeSluttAlder.maaneder
      : 11

  return periodEndMonth - periodStartMonth + 1
}

export const processPensjonsavtalerArray = (
  xAxisStartAar: number, // uttaksaar, minus 1
  xAxisLength: number,
  privatePensjonsavtaler: Pensjonsavtale[],
  offentligTpUtbetalingsperioder: UtbetalingsperiodeWithoutGrad[]
): number[] => {
  const sluttAlder = xAxisStartAar + xAxisLength - 1
  const result = new Array(sluttAlder - xAxisStartAar + 1).fill(0)

  const samledeUtbetalingsperioder = [
    ...privatePensjonsavtaler.flatMap((avtale) => avtale.utbetalingsperioder),
    ...offentligTpUtbetalingsperioder,
  ]

  samledeUtbetalingsperioder.forEach((utbetalingsperiode) => {
    const avtaleStartYear = Math.max(
      xAxisStartAar,
      utbetalingsperiode.startAlder.aar
    )
    const avtaleEndYear = utbetalingsperiode.sluttAlder
      ? Math.min(sluttAlder, utbetalingsperiode.sluttAlder.aar)
      : sluttAlder

    for (let year = avtaleStartYear; year <= avtaleEndYear; year++) {
      if (year >= xAxisStartAar) {
        const antallMaanederMedPensjon = getAntallMaanederMedPensjon(
          year,
          utbetalingsperiode.startAlder,
          utbetalingsperiode.sluttAlder
        )

        const allocatedAmount =
          (utbetalingsperiode.aarligUtbetaling *
            Math.max(0, antallMaanederMedPensjon)) /
          12
        result[year - xAxisStartAar] += allocatedAmount
      }
    }
  })
  return result
}

export const generateXAxis = (
  xAxisStartAar: number,
  isEndring: boolean,
  privatePensjonsavtaler: Pensjonsavtale[],
  offentligTpUtbetalingsperioder: UtbetalingsperiodeWithoutGrad[],
  setIsPensjonsavtaleFlagVisible: React.Dispatch<React.SetStateAction<boolean>>
) => {
  let sluttAar = MAX_UTTAKSALDER
  let hasAvtaleBeforeStartAlder = false

  privatePensjonsavtaler.forEach((avtale: Pensjonsavtale) => {
    if (avtale.sluttAar && avtale.sluttAar > sluttAar) {
      sluttAar = avtale.sluttAar
    }
    if (
      !hasAvtaleBeforeStartAlder &&
      avtale.startAar &&
      avtale.startAar < xAxisStartAar
    ) {
      hasAvtaleBeforeStartAlder = true
    }
  })

  offentligTpUtbetalingsperioder.forEach(
    (utbetalingsperiode: UtbetalingsperiodeWithoutGrad) => {
      if (utbetalingsperiode.startAlder.aar > sluttAar) {
        sluttAar = utbetalingsperiode.startAlder.aar
      }
      if (
        utbetalingsperiode.sluttAlder &&
        utbetalingsperiode.sluttAlder.aar > sluttAar
      ) {
        sluttAar = utbetalingsperiode.sluttAlder.aar
      }
      if (
        !hasAvtaleBeforeStartAlder &&
        utbetalingsperiode.startAlder.aar < xAxisStartAar
      ) {
        hasAvtaleBeforeStartAlder = true
      }
    }
  )

  const alderArray: string[] = []
  for (let i = xAxisStartAar; i <= sluttAar + 1; i++) {
    if (!isEndring && i === xAxisStartAar) {
      alderArray.push((i - 1).toString())
    }

    if (i === sluttAar + 1) {
      alderArray.push(`${i - 1}+`.toString())
    } else {
      alderArray.push(i.toString())
    }
  }
  setIsPensjonsavtaleFlagVisible(hasAvtaleBeforeStartAlder)
  return alderArray
}

export function getTooltipTitle(
  antallAar: string,
  hasInntekt: boolean,
  hasPensjon: boolean,
  intl: IntlShape
): string {
  if (hasInntekt && hasPensjon) {
    return `${intl.formatMessage({
      id: 'beregning.highcharts.tooltip.inntekt_og_pensjon',
    })} ${antallAar} ${intl.formatMessage({
      id: 'alder.aar',
    })}`
  } else if (hasInntekt && !hasPensjon) {
    return `${intl.formatMessage({
      id: 'beregning.highcharts.tooltip.inntekt',
    })} ${antallAar} ${intl.formatMessage({
      id: 'alder.aar',
    })}`
  } else {
    return `${intl.formatMessage({
      id: 'beregning.highcharts.tooltip.pensjon',
    })} ${antallAar} ${intl.formatMessage({
      id: 'alder.aar',
    })}`
  }
}

export const onVisFaerreAarClick = () => {
  const el = document.querySelector(highchartsScrollingSelector)
  if (el) {
    ;(el as HTMLElement).scrollLeft -= 50
  }
}

export const onVisFlereAarClick = () => {
  const el = document.querySelector(highchartsScrollingSelector)
  if (el) {
    ;(el as HTMLElement).scrollLeft += 50
  }
}

export const getHoverColor = (previousColor: string): string => {
  switch (previousColor) {
    case SERIES_COLORS.SERIE_COLOR_INNTEKT: {
      return SERIES_COLORS.SERIE_COLOR_FADED_INNTEKT
    }
    case SERIES_COLORS.SERIE_COLOR_AFP: {
      return SERIES_COLORS.SERIE_COLOR_FADED_AFP
    }
    case SERIES_COLORS.SERIE_COLOR_TP: {
      return SERIES_COLORS.SERIE_COLOR_FADED_TP
    }
    case SERIES_COLORS.SERIE_COLOR_ALDERSPENSJON: {
      return SERIES_COLORS.SERIE_COLOR_FADED_ALDERSPENSJON
    }
    default: {
      return ''
    }
  }
}

export const getNormalColor = (previousColor: string): string => {
  switch (previousColor) {
    case SERIES_COLORS.SERIE_COLOR_FADED_INNTEKT: {
      return SERIES_COLORS.SERIE_COLOR_INNTEKT
    }
    case SERIES_COLORS.SERIE_COLOR_FADED_AFP: {
      return SERIES_COLORS.SERIE_COLOR_AFP
    }
    case SERIES_COLORS.SERIE_COLOR_FADED_TP: {
      return SERIES_COLORS.SERIE_COLOR_TP
    }
    case SERIES_COLORS.SERIE_COLOR_FADED_ALDERSPENSJON: {
      return SERIES_COLORS.SERIE_COLOR_ALDERSPENSJON
    }
    default: {
      return previousColor
    }
  }
}

export function resetColumnColors(chart: Chart): void {
  if (chart.series) {
    chart.series.forEach(function (serie: Series) {
      serie.data.forEach(function (point: Point) {
        const color = getNormalColor(point.color as string)
        if (point.color !== color) {
          point.update({ color: getNormalColor(point.color as string) }, false)
        }
      })
    })
  }

  if ((chart.xAxis[0] as ExtendedAxis).labelGroup) {
    ;(chart.xAxis[0] as ExtendedAxis).labelGroup.element.childNodes.forEach(
      function (label: HTMLElement) {
        label.style.fontWeight = 'normal'
      }
    )
  }
  chart?.tooltip.update({
    enabled: false,
  })
  chart.redraw()
}

export function handleChartScroll(
  event: Event,
  args: {
    chart?: Chart
    scrollPosition?: number
  }
) {
  if (event.currentTarget) {
    const { chart, scrollPosition } = args
    const el = event.currentTarget as HighchartsScrollingHTMLDivElement
    const elementScrollPosition = el.scrollLeft
    if (chart && scrollPosition !== el.scrollLeft) {
      resetColumnColors(chart)
    }

    const isRightButtonAvailable = el.scrollWidth > el.offsetWidth

    if (elementScrollPosition === 0) {
      el.handleButtonVisibility.showRightButton(isRightButtonAvailable)
      el.handleButtonVisibility.showLeftButton(false)
    } else if (elementScrollPosition + el.offsetWidth === el.scrollWidth) {
      el.handleButtonVisibility.showRightButton(false)
      el.handleButtonVisibility.showLeftButton(true)
    } else {
      el.handleButtonVisibility.showRightButton(isRightButtonAvailable)
      el.handleButtonVisibility.showLeftButton(true)
    }
  }
}
