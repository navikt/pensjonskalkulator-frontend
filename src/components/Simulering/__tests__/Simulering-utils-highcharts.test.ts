import { IntlShape } from 'react-intl'

import {
  AxisLabelsFormatterContextObject,
  TooltipFormatterContextObject,
  Chart,
  Point,
} from 'highcharts'
import { describe, expect, it, vi } from 'vitest'

import { SERIES_DEFAULT } from '../constants'
import { onVisFlereAarClick } from '../utils'
import {
  ExtendedAxis,
  ExtendedPoint,
  getChartOptions,
  labelFormatterDesktop,
  labelFormatterMobile,
  onPointClick,
  onPointUnclick,
  tooltipFormatter,
} from '../utils-highcharts'

import { getChartMock } from './chart-mock'

import globalClassNames from './Simulering.module.scss'

describe('Simulering-utils-highcharts', () => {
  describe('labelFormatterMobile', () => {
    it('returnerer riktig streng når verdien er under 1000', () => {
      const a = labelFormatterMobile.bind({
        value: 300,
      } as AxisLabelsFormatterContextObject)
      expect(a()).toEqual('300')
    })
    it('returnerer riktig streng når verdien er lik 1000', () => {
      const a = labelFormatterMobile.bind({
        value: 1000,
      } as AxisLabelsFormatterContextObject)
      expect(a()).toEqual('1000')
    })
    it('returnerer riktig streng når verdien er over 1000', () => {
      const a = labelFormatterMobile.bind({
        value: 1000000,
      } as AxisLabelsFormatterContextObject)
      expect(a()).toEqual('1000')
    })
    it('returnerer riktig når verdien er av type streng', () => {
      const a = labelFormatterMobile.bind({
        value: '1000000',
      } as AxisLabelsFormatterContextObject)
      expect(a()).toEqual('1000')
    })
  })
  describe('labelFormatterDesktop', () => {
    it('returnerer riktig streng når verdien er under 1000', () => {
      const a = labelFormatterDesktop.bind({
        value: 300,
      } as AxisLabelsFormatterContextObject)
      expect(a()).toEqual('300')
    })
    it('returnerer riktig streng når verdien er lik 1000', () => {
      const a = labelFormatterDesktop.bind({
        value: 1000,
      } as AxisLabelsFormatterContextObject)
      expect(a()).toEqual('1 000')
    })
    it('returnerer riktig streng når verdien er over 1000', () => {
      const a = labelFormatterDesktop.bind({
        value: 1000000,
      } as AxisLabelsFormatterContextObject)
      expect(a()).toEqual('1 000 000')
    })
    it('returnerer riktig når verdien er av type streng', () => {
      const a = labelFormatterDesktop.bind({
        value: '1000000',
      } as AxisLabelsFormatterContextObject)
      expect(a()).toEqual('1 000 000')
    })
  })
  describe('tooltipFormatter', () => {
    it('returnerer formatert tooltip med riktig data og stiler for begge serier', () => {
      const stylesMock = {
        tooltipLine: 'tooltipLine',
        tooltipTable: 'tooltipTable',
        tooltipTableHeaderCell: 'tooltipTableHeaderCell',
        tooltipTableHeaderCell__left: 'tooltipTableHeaderCell__left',
        tooltipTableHeaderCell__right: 'tooltipTableHeaderCell__right',
        tooltipTableCell: 'tooltipTableCell',
        tooltipTableCell__right: 'tooltipTableCell__right',
        tooltipTableCellDot: 'tooltipTableCellDot',
      } as Partial<typeof globalClassNames>

      const alder = 65
      const total = 800000
      const colorSerie1 = 'lime'
      const colorSerie2 = 'salmon'
      const plotY = 68
      const pointSumSerie1 = 200000
      const pointSumSerie2 = 350000

      const beregnetLinePosition = 'top: 135px; left: 34px; height: 34px'
      const beregnetLinePositionAfterScroll =
        'top: 135px; left: -16px; height: 34px'

      const simplePoint = {
        y: pointSumSerie1,
        percentage: 20,
        total,
        series: {
          name: SERIES_DEFAULT.SERIE_INNTEKT.name,
          color: colorSerie1,
          yAxis: { height: 171 } as ExtendedAxis,
        },
      }

      const context = {
        point: { plotY },
        x: alder,
        points: [
          {
            y: pointSumSerie1,
            percentage: 20,
            total,
            series: {
              name: SERIES_DEFAULT.SERIE_INNTEKT.name,
              color: colorSerie1,
              chart: {
                chartHeight: 400,
                plotLeft: 35,
                series: [
                  {
                    name: SERIES_DEFAULT.SERIE_INNTEKT.name,
                    color: colorSerie1,
                    chart: { chartHeight: 400, plotLeft: 35 },
                    data: [
                      {
                        ...simplePoint,
                      },
                      {
                        ...simplePoint,
                        y: pointSumSerie2,
                        series: {
                          ...simplePoint.series,
                          name: SERIES_DEFAULT.SERIE_ALDERSPENSJON.name,
                          color: colorSerie2,
                        },
                      },
                    ],
                  },
                  {
                    name: SERIES_DEFAULT.SERIE_ALDERSPENSJON.name,
                    color: colorSerie2,
                    chart: { chartHeight: 400, plotLeft: 35 },
                    data: [
                      {
                        ...simplePoint,
                      },
                      {
                        ...simplePoint,
                        y: pointSumSerie2,
                        series: {
                          ...simplePoint.series,
                          name: SERIES_DEFAULT.SERIE_ALDERSPENSJON.name,
                          color: colorSerie2,
                        },
                      },
                    ],
                  },
                ],
              },
            },
            point: {
              plotX: 129,
              series: {
                data: ['70', '71', '72', '73', '74', '75', '76', '77+'],
              },
            } as ExtendedPoint,
          },
        ],
      }
      const tooltipMarkup = tooltipFormatter(
        context as unknown as TooltipFormatterContextObject,
        stylesMock,
        {
          formatMessage: (s: { id: string }) => s.id,
        } as unknown as IntlShape
      )
      expect(tooltipMarkup).toContain('800 000 kr')
      expect(tooltipMarkup).toContain(
        `beregning.highcharts.tooltip.inntekt_og_pensjon 65 alder.aar`
      )
      expect(tooltipMarkup).toContain(SERIES_DEFAULT.SERIE_INNTEKT.name)
      expect(tooltipMarkup).toContain(SERIES_DEFAULT.SERIE_ALDERSPENSJON.name)
      expect(tooltipMarkup).toContain(`backgroundColor:${colorSerie1}`)
      expect(tooltipMarkup).toContain(`backgroundColor:${colorSerie2}`)
      expect(tooltipMarkup).toContain('200 000 kr')
      expect(tooltipMarkup).toContain('350 000 kr')
      expect(tooltipMarkup).toContain(beregnetLinePosition)

      const div = document.createElement('div')
      div.innerHTML = '<div class="highcharts-scrolling">SPAN</div>'
      document.body.appendChild(div)

      onVisFlereAarClick()

      const tooltipMarkupAfterScroll = tooltipFormatter(
        context as unknown as TooltipFormatterContextObject,
        stylesMock,
        {
          formatMessage: (s: { id: string }) => s.id,
        } as unknown as IntlShape
      )
      expect(tooltipMarkupAfterScroll).toContain(
        beregnetLinePositionAfterScroll
      )
    })
  })
  describe('onPointClick og onPointUnclick', () => {
    describe('onPointClick', () => {
      it('oppdaterer fargen på kolonnen som er valgt og de som ikke er det samt label i xAxis', () => {
        const redrawMock = vi.fn()
        const pointUpdateMock = vi.fn()
        const tooltipUpdateMock = vi.fn()
        const tooltipRefreshMock = vi.fn()

        const chart = getChartMock(
          pointUpdateMock,
          tooltipUpdateMock,
          tooltipRefreshMock
        )

        const point = {
          index: 0,
          series: {
            chart: {
              ...chart,
              redraw: redrawMock,
            } as unknown as Chart,
          },
        } as Point
        onPointClick.call(point)
        expect(tooltipUpdateMock).toHaveBeenCalled()
        expect(pointUpdateMock).toHaveBeenCalledTimes(3)
        expect(pointUpdateMock.mock.calls).toEqual([
          [{ color: 'var(--a-deepblue-200)' }, false],
          [{ color: 'var(--a-data-surface-5-subtle)' }, false],
          [{ color: 'var(--a-purple-200)' }, false],
        ])
        expect(redrawMock).toHaveBeenCalledOnce()
        expect(tooltipRefreshMock).toHaveBeenCalledOnce()
      })

      describe('onPointUnclick', () => {
        const pointUpdateMock = vi.fn()
        const tooltipUpdateMock = vi.fn()
        const tooltipRefreshMock = vi.fn()

        const chart = getChartMock(
          pointUpdateMock,
          tooltipUpdateMock,
          tooltipRefreshMock
        )

        it('gjør ingenting når chart ikke er klar', () => {
          expect(onPointUnclick({} as MouseEvent, undefined)).toBe(undefined)
        })

        it('gjør ingenting når brukeren klikker på et chart point', () => {
          const event = { chartX: 123, point: {} as unknown as Point }
          expect(
            onPointUnclick(
              event as unknown as MouseEvent,
              { ...chart } as unknown as Chart
            )
          ).toBe(undefined)
        })

        it('kaller resetColumnColors og nullstiller fargen på alle kolonnene når brukeren klikker utenfor en chart point', () => {
          const redrawMock = vi.fn()
          vi.useFakeTimers()
          const chartWithSelection = {
            ...chart,
            redraw: redrawMock,
            tooltip: {
              ...chart.tooltip,
              isHidden: false,
            },
          } as unknown as Chart
          onPointUnclick(
            { chartX: 123 } as unknown as MouseEvent,
            chartWithSelection
          )
          vi.advanceTimersByTime(150)
          expect(pointUpdateMock).toHaveBeenCalledTimes(3)
          expect(pointUpdateMock.mock.calls).toEqual([
            [{ color: 'var(--a-deepblue-500)' }, false],
            [{ color: 'var(--a-data-surface-5)' }, false],
            [{ color: 'var(--a-purple-400)' }, false],
          ])
          expect(redrawMock).toHaveBeenCalledOnce()
          expect(tooltipUpdateMock).toHaveBeenCalledOnce()
        })

        it('kaller resetColumnColors og nullstiller fargen på alle kolonnene når brukeren klikker utenfor plot area og at tooltip er skjult', () => {
          const redrawMock = vi.fn()
          vi.useFakeTimers()
          const chartWithSelection = {
            ...chart,
            redraw: redrawMock,
            tooltip: {
              ...chart.tooltip,
              isHidden: true,
            },
          } as unknown as Chart
          onPointUnclick({} as unknown as MouseEvent, chartWithSelection)
          vi.advanceTimersByTime(150)
          expect(pointUpdateMock).toHaveBeenCalledTimes(6)
          expect(pointUpdateMock.mock.calls).toEqual([
            [
              {
                color: 'var(--a-deepblue-500)',
              },
              false,
            ],
            [
              {
                color: 'var(--a-data-surface-5)',
              },
              false,
            ],
            [
              {
                color: 'var(--a-purple-400)',
              },
              false,
            ],
            [
              {
                color: 'var(--a-deepblue-500)',
              },
              false,
            ],
            [
              {
                color: 'var(--a-data-surface-5)',
              },
              false,
            ],
            [
              {
                color: 'var(--a-purple-400)',
              },
              false,
            ],
          ])
          expect(redrawMock).toHaveBeenCalledOnce()
          expect(tooltipUpdateMock).toHaveBeenCalledTimes(2)
        })
      })
    })

    describe('getChartOptions', () => {
      it('returnerer riktig default options', () => {
        const options = getChartOptions(
          {} as Partial<typeof globalClassNames>,
          vi.fn(),
          vi.fn(),
          {
            formatMessage: (s: { id: string }) => s.id,
          } as unknown as IntlShape
        )
        expect(options).toMatchSnapshot()
      })
    })
  })
})
