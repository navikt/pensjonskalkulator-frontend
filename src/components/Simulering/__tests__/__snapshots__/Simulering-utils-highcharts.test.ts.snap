// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Simulering-utils-highcharts > onPointClick og onPointUnclick > getChartOptions > returnerer riktig default options 1`] = `
{
  "chart": {
    "animation": false,
    "events": {
      "render": [Function],
    },
    "scrollablePlotArea": {
      "minWidth": 750,
      "scrollPositionX": 0,
    },
    "spacingBottom": 0,
    "spacingLeft": 0,
    "spacingTop": 0,
    "style": {
      "zIndex": 1,
    },
    "type": "column",
  },
  "credits": {
    "enabled": false,
  },
  "legend": {
    "accessibility": {
      "enabled": true,
      "keyboardNavigation": {
        "enabled": false,
      },
    },
    "align": "left",
    "events": {
      "itemClick": [Function],
    },
    "itemDistance": 24,
    "itemHoverStyle": {
      "color": "#000000",
    },
    "itemMarginBottom": 5,
    "itemStyle": {
      "color": "var(--a-text-default)",
      "cursor": "default",
      "fontFamily": "var(--a-font-family)",
      "fontSize": "14px",
      "fontWeight": "regular",
      "zIndex": 0,
    },
    "layout": "horizontal",
    "margin": 0,
    "padding": 0,
    "useHTML": true,
    "verticalAlign": "bottom",
    "x": 0,
    "y": -25,
  },
  "plotOptions": {
    "column": {
      "point": {
        "events": {
          "click": [Function],
          "mouseOut": [Function],
          "mouseOver": [Function],
        },
      },
    },
    "series": {
      "animation": false,
      "events": {
        "legendItemClick": [Function],
      },
      "stacking": "normal",
      "states": {
        "inactive": {
          "enabled": false,
        },
      },
      "stickyTracking": false,
    },
  },
  "responsive": {
    "rules": [
      {
        "chartOptions": {
          "legend": {
            "itemDistance": 12,
            "margin": 50,
            "verticalAlign": "top",
          },
          "tooltip": {
            "outside": true,
            "positioner": [Function],
          },
          "xAxis": {
            "labels": {
              "style": {
                "fontSize": "var(--a-font-size-small)",
              },
            },
            "title": {
              "text": "",
            },
          },
          "yAxis": {
            "labels": {
              "formatter": [Function],
              "style": {
                "backgroundColor": "transparent",
                "fontSize": "var(--a-font-size-small)",
              },
              "x": 0,
            },
            "offset": 28,
            "title": {
              "margin": -75,
              "style": {
                "fontSize": "var(--a-font-size-small)",
              },
              "text": "beregning.highcharts.yaxis.mobile",
              "x": -73,
              "y": -22,
            },
          },
        },
        "condition": {
          "callback": [Function],
        },
      },
    ],
  },
  "series": [],
  "title": {
    "align": "left",
    "margin": 20,
    "style": {
      "opacity": 0,
    },
    "text": "-",
    "y": 0,
  },
  "tooltip": {
    "animation": false,
    "className": undefined,
    "followTouchMove": false,
    "formatter": [Function],
    "hideDelay": 9000000000,
    "outside": false,
    "padding": 0,
    "positioner": [Function],
    "shadow": false,
    "shared": true,
    "useHTML": true,
  },
  "xAxis": {
    "categories": [],
    "labels": {
      "formatter": [Function],
      "style": {
        "color": "var(--a-text-subtle)",
        "fontFamily": "var(--a-font-family)",
        "fontSize": "var(--a-font-size-medium)",
      },
      "y": 20,
    },
    "lineColor": "var(--a-text-subtle)",
    "title": {
      "align": "high",
      "style": {
        "color": "var(--a-text-subtle)",
        "fontFamily": "var(--a-font-family)",
        "fontSize": "var(--a-font-size-medium)",
      },
      "text": "beregning.highcharts.xaxis",
      "y": -5,
    },
  },
  "yAxis": {
    "allowDecimals": false,
    "gridLineColor": "var(--a-gray-400)",
    "gridLineWidth": 1,
    "labels": {
      "align": "left",
      "formatter": [Function],
      "style": {
        "color": "var(--a-text-subtle)",
        "fontFamily": "var(--a-font-family)",
        "fontSize": "var(--a-font-size-medium)",
        "paddingRight": "var(--a-spacing-3)",
      },
      "useHTML": true,
      "x": -55,
    },
    "min": 0,
    "minorGridLineWidth": 0,
    "minorTickInterval": 50000,
    "offset": 10,
    "tickInterval": 100000,
    "title": {
      "align": "high",
      "rotation": 0,
      "style": {
        "fontFamily": "var(--a-font-family)",
        "fontSize": "var(--a-font-size-medium)",
        "zIndex": 0,
      },
      "text": "beregning.highcharts.yaxis",
      "textAlign": "left",
      "x": -44,
      "y": -20,
    },
  },
}
`;
